File: python/tree/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getKth(self,root,k):
        if not root:
            return 0,0

        val,left = self.getKth(root.left,k)
        if left + 1 > k:
            return val,left
        elif left + 1 == k:
            return root.val,k

        val,right =  self.getKth(root.right,k - left - 1)
        return val, left + right + 1

    def kthSmallest(self, root, k):
        val, count = self.getKth(root,k)
        return val

File: c/tree/kth-smallest-element-in-a-bst.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int getKth(struct TreeNode *root, int k, int *val)
{
    if(root == NULL)
        return 0;

    int left = getKth(root->left,k, val);
    if(left + 1 > k)
        return left;
    else if(left + 1 == k){
        *val = root->val;
        return left + 1;
    }

    return left + 1 + getKth(root->right, k - left - 1, val);
}

int kthSmallest(struct TreeNode* root, int k) {
    int val;
    getKth(root,k,&val);

    return val;
}
